src/orchard/util/cljs.clj

```diff
@@ -0,0 +1,101 @@
+(ns orchard.util.cljs)
+
+(defn try-piggieback
+  "If piggieback is loaded, returns `#'cemerick.piggieback/wrap-cljs-repl`, or
+  false otherwise."
+  []
+  (try
+    (require 'cemerick.piggieback)
+    (resolve 'cemerick.piggieback/wrap-cljs-repl)
+    (catch Exception _)))
+
+;; TODO: warn on usage of piggieback 0.1.x?
+(defn piggieback-0-2+?
+  "Returns true if piggieback 0.2.x is loaded, or false otherwise."
+  []
+  (boolean
+   (try
+     (require 'cemerick.piggieback)
+     (resolve 'cemerick.piggieback/*cljs-compiler-env*)
+     (catch Exception _))))
+
+(defn- maybe-piggieback
+  [descriptor descriptor-key]
+  (if-let [piggieback (try-piggieback)]
+    (update-in descriptor [descriptor-key] #(set (conj % piggieback)))
+    descriptor))
+
+(defn expects-piggieback
+  "If piggieback is loaded, returns the descriptor with piggieback's
+  `wrap-cljs-repl` handler assoc'd into its `:expects` set."
+  [descriptor]
+  (maybe-piggieback descriptor :expects))
+
+(defn requires-piggieback
+  "If piggieback is loaded, returns the descriptor with piggieback's
+  `wrap-cljs-repl` handler assoc'd into its `:requires` set."
+  [descriptor]
+  (maybe-piggieback descriptor :requires))
+
+(defn- cljs-env-path
+  "Returns the path in the session map for the ClojureScript compiler
+  environment used by piggieback."
+  []
+  (if (piggieback-0-2+?)
+    [(resolve 'cemerick.piggieback/*cljs-compiler-env*)]
+    [(resolve 'cemerick.piggieback/*cljs-repl-env*) :cljs.env/compiler]))
+
+(defn- maybe-deref
+  [x]
+  (if (instance? clojure.lang.IDeref x) @x x))
+
+(defn- grab-cljs-env*
+  [msg]
+  (let [path (cljs-env-path)]
+    (some-> msg
+            :session
+            maybe-deref
+            (get-in path))))
+
+(defn grab-cljs-env
+  "If piggieback is active, returns the ClojureScript compiler environment for
+  the running REPL."
+  [msg]
+  (maybe-deref (grab-cljs-env* msg)))
```

Comment:
Seems that this is making use of an nREPL message. I've yet to figure out how to untangle the CLJS stuff from the nREPL messaging--any pointers would be appreciated.

Reply:
That's why I left this in `cider-nrepl` for now - wasn't quite clear to me what's the best way to decouple things from Piggieback, and there's also `cljs-tooling` to consider, which should likely be integrated into `orchard` itself using `cljc`. Technically the completion needs just the cljs compiler state to work, but everything piggieback related has to remain in `cider-nrepl`, as it's nREPL specific.